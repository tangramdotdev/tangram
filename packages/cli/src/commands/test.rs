use crate::Cli;
use tangram_client as tg;
use tangram_error::Result;

/// Build the target named "test" from the specified package.
#[derive(Debug, clap::Args)]
#[command(verbatim_doc_comment)]
pub struct Args {
	/// If this flag is set, then the command will exit immediately instead of waiting for the build's output.
	#[arg(short, long)]
	pub detach: bool,

	/// The path to the executable in the artifact to run.
	#[arg(long)]
	pub executable_path: Option<tg::Path>,

	/// If this flag is set, the package's lockfile will not be updated.
	#[arg(long)]
	pub locked: bool,

	/// The package to build.
	#[arg(short, long, default_value = ".")]
	pub package: tg::Dependency,

	/// The retry strategy to use.
	#[arg(long, default_value_t)]
	pub retry: tg::build::Retry,
}

impl Cli {
	pub async fn command_test(&self, args: Args) -> Result<()> {
		// Create the build args.
		let args = super::build::Args {
			detach: args.detach,
			locked: args.locked,
			no_tui: false,
			output: None,
			package: args.package,
			retry: args.retry,
			target: "test".to_owned(),
		};

		// Build!
		self.command_build(args).await?;

		Ok(())
	}
}
