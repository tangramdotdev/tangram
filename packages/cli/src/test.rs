use crate::Config;
use futures::{Future, FutureExt as _};
use std::{panic::AssertUnwindSafe, sync::Arc};
use tangram_client as tg;
use tangram_temp::Temp;
use url::Url;

pub async fn test<F, Fut>(tg: &'static str, f: F)
where
	F: FnOnce(Arc<tokio::sync::Mutex<Context>>) -> Fut,
	Fut: Future<Output = ()> + Send,
{
	// Create the context.
	let context = Arc::new(tokio::sync::Mutex::new(Context::new(tg)));

	// Run the test and catch a panic if one occurs.
	let result = AssertUnwindSafe(f(context.clone())).catch_unwind().await;

	// Handle the result.
	let context = context.lock().await;
	match result {
		Ok(()) => {
			// If there was no panic, then gracefully shutdown the servers.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGINT) };
				let result = process.wait().await;
				if let Ok(status) = result {
					if !status.success() {}
				}
			}
		},
		Err(error) => {
			// If there was a panic, then forcefully shutdown the servers and resume unwinding.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGKILL) };
				process.wait().await.ok();
			}
			std::panic::resume_unwind(error);
		},
	}
}

pub struct Context {
	servers: Vec<Arc<Server>>,
	tg: &'static str,
}

pub struct Server {
	config: Config,
	process: tokio::sync::Mutex<tokio::process::Child>,
	temp: Arc<Temp>,
	tg: &'static str,
}

impl Context {
	#[must_use]
	pub fn new(tg: &'static str) -> Self {
		Self {
			servers: Vec::new(),
			tg,
		}
	}

	pub async fn spawn_server(&mut self) -> tg::Result<Arc<Server>> {
		let server = Server::new(self.tg).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_config(&mut self, config: Config) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_temp_and_config(
		&mut self,
		config: Config,
	) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}
}

impl Server {
	async fn new(tg: &'static str) -> tg::Result<Self> {
		let config = Config::default();
		Self::with_config(tg, config).await
	}

	async fn with_config(tg: &'static str, config: Config) -> tg::Result<Self> {
		// Create a temp and create the config and data paths.
		let temp = Arc::new(Temp::new());
		tokio::fs::create_dir_all(temp.path())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the directory"))?;
		let config_path = temp.path().join(".config/tangram/config.json");
		tokio::fs::create_dir_all(config_path.parent().unwrap())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the config directory"))?;
		let config_json = serde_json::to_vec_pretty(&config).unwrap();
		tokio::fs::write(&config_path, config_json)
			.await
			.map_err(|source| tg::error!(!source, "failed to write the config"))?;
		let data_path = temp.path().join(".tangram");
		tokio::fs::create_dir_all(&data_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to create the data directory"))?;

		// Create the command.
		let log_path = temp.path().join(".tangram/log");
		let stdout = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let stderr = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let mut command = tokio::process::Command::new(tg);
		command.stdout(stdout);
		command.stderr(stderr);
		command.arg("--config");
		command.arg(&config_path);
		command.arg("--path");
		command.arg(data_path);
		command.arg("serve");

		// Spawn the process.
		let process = command.spawn().unwrap();
		let process = tokio::sync::Mutex::new(process);

		// Create the server.
		let server = Self {
			config,
			process,
			temp,
			tg,
		};

		Ok(server)
	}

	#[must_use]
	pub fn tg(&self) -> tokio::process::Command {
		let mut command = tokio::process::Command::new(self.tg);
		let config_path = self.temp.path().join(".config/tangram/config.json");
		let data_path = self.temp.path().join(".tangram");
		command
			.arg("--config")
			.arg(config_path)
			.arg("--path")
			.arg(data_path)
			.arg("--mode")
			.arg("client");
		command
	}

	#[must_use]
	pub fn config(&self) -> &Config {
		&self.config
	}

	#[must_use]
	pub fn process(&self) -> &tokio::sync::Mutex<tokio::process::Child> {
		&self.process
	}

	#[must_use]
	pub fn url(&self) -> Url {
		let path = self.temp.path().join(".tangram/socket");
		let path = path.to_str().unwrap();
		let path = urlencoding::encode(path);
		format!("http+unix://{path}").parse().unwrap()
	}

	pub async fn print_log(&self) -> tg::Result<()> {
		let log_path = self.temp.join(".tangram/log");
		let mut log_file = tokio::fs::File::open(log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?;
		let mut stderr = tokio::io::stderr();
		tokio::io::copy(&mut log_file, &mut stderr).await.ok();
		Ok(())
	}
}

#[macro_export]
macro_rules! assert_output_success {
	($output:expr) => {
		let output = &$output;
		if !output.status.success() {
			let mut stderr = tokio::io::stderr();
			stderr.write_all(&output.stderr).await.unwrap();
			stderr.flush().await.unwrap();
		}
		assert!(output.status.success());
	};
}
pub use assert_output_success;
