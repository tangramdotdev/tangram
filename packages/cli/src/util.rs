use console::style;
use either::Either;
use tangram_client as tg;
use tangram_error::Error;

pub fn build_or_object_id(s: &str) -> Result<Either<tg::build::Id, tg::object::Id>, String> {
	if let Ok(value) = s.parse() {
		return Ok(Either::Left(value));
	}
	if let Ok(value) = s.parse() {
		return Ok(Either::Right(value));
	}
	Err("Failed to parse.".to_string())
}

pub fn print_error(mut error: &tangram_error::Error) {
	eprintln!("{}:", style("Error").red());
	let mut first = true;
	loop {
		if !first {
			eprintln!();
		}
		first = false;
		eprint!("{}", style("->").red());
		let Error {
			message,
			location,
			stack,
			source,
			values,
		} = error;
		eprint!(" {message}");
		if let Some(location) = &location {
			let location = style(location).dim().white();
			eprint!(" {location}");
		}
		for (name, value) in values {
			eprintln!();
			let name = style(name).blue();
			let value = style(value).green();
			eprint!("   {name} = {value}");
		}
		for location in stack.iter().flatten() {
			eprintln!();
			let location = style(location).dim().white();
			eprint!(" {location}");
		}
		if let Some(source) = &source {
			error = source;
		} else {
			break;
		}
	}
	eprintln!();
}
