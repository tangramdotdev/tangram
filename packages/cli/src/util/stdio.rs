use bytes::Bytes;
use futures::{Stream, TryStreamExt as _};
use std::io::Read as _;
use tangram_client as tg;
use tokio::io::AsyncBufRead;
use tokio_stream::wrappers::ReceiverStream;
use tokio_util::io::StreamReader;

pub(crate) fn stdin() -> impl AsyncBufRead + Send + 'static {
	StreamReader::new(stdin_stream().map_err(std::io::Error::other))
}

pub(crate) fn stdin_stream() -> impl Stream<Item = tg::Result<Bytes>> + Send + 'static {
	let (send, recv) = tokio::sync::mpsc::channel(1);
	std::thread::spawn(move || {
		let mut stdin = std::io::stdin();
		loop {
			let mut buf = vec![0u8; 4096];
			let result = match stdin.read(&mut buf) {
				Ok(0) => break,
				Ok(n) => Ok(Bytes::copy_from_slice(&buf[0..n])),
				Err(source) => Err(tg::error!(!source, "failed to read stdin")),
			};
			let result = send.blocking_send(result);
			if result.is_err() {
				break;
			}
		}
	});
	ReceiverStream::new(recv)
}
