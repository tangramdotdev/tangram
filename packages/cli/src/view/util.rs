use ratatui::{prelude::*, widgets::Block};

pub fn render_block_and_get_area(title: &str, focused: bool, area: Rect, buf: &mut Buffer) -> Rect {
	let block = Block::bordered()
		.title(title)
		.border_style(Style::default().fg(if focused { Color::Blue } else { Color::White }));
	block.render(area, buf);
	Layout::default()
		.constraints([Constraint::Percentage(100)])
		.margin(1)
		.split(area)
		.first()
		.copied()
		.unwrap_or(area)
}
