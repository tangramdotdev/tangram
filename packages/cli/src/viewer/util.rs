use {
	futures::TryStreamExt as _,
	std::{io::Write as _, pin::pin},
	tangram_client::{self as tg, handle::Ext as _},
};

const BLOB_LENGTH_LIMIT: u64 = 1 << 20;

pub async fn format_blob<H: tg::Handle>(handle: &H, blob: &tg::Blob) -> tg::Result<String> {
	let length = blob.length(handle).await?;
	if length > BLOB_LENGTH_LIMIT {
		return Err(tg::error!("cannot view blobs larger than 1Mib"));
	}
	let mut contents = Vec::new();
	let arg = tg::read::Arg {
		blob: blob.id(),
		options: tg::read::Options {
			length: Some(BLOB_LENGTH_LIMIT),
			..tg::read::Options::default()
		},
	};
	let stream = handle
		.try_read(arg)
		.await?
		.ok_or_else(|| tg::error!("failed to read the blob"))?;
	let mut stream = pin!(stream);
	let mut is_utf8 = true;
	while let Some(chunk) = stream.try_next().await? {
		if let (true, Ok(chunk)) = (is_utf8, std::str::from_utf8(&chunk.bytes)) {
			contents.extend_from_slice(chunk.as_bytes());
			continue;
		}
		is_utf8 = false;
		contents.reserve(chunk.bytes.len());
		for byte in chunk.bytes {
			if byte.is_ascii_graphic() {
				contents.push(byte);
			} else {
				write!(&mut contents, " {byte:20x}").unwrap();
			}
		}
	}
	Ok(String::from_utf8(contents).unwrap())
}
