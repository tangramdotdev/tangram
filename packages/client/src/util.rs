pub mod arc;
pub mod fs;
pub mod http;
pub mod serde;
