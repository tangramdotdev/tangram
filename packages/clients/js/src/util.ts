import type * as tg from "./index.ts";

export type MaybePromise<T> = T | Promise<T>;

export type MaybeMutation<T extends tg.Value = tg.Value> = T | tg.Mutation<T>;

export type MutationMap<
	T extends { [key: string]: tg.Value } = { [key: string]: tg.Value },
> = {
	[K in keyof T]?: tg.Mutation<T[K]>;
};

export type MaybeMutationMap<
	T extends { [key: string]: tg.Value } = { [key: string]: tg.Value },
> = {
	[K in keyof T]?: tg.MaybeMutation<T[K]>;
};

export type ValueOrMaybeMutationMap<T extends tg.Value = tg.Value> = T extends
	| undefined
	| boolean
	| number
	| string
	| tg.Object
	| Uint8Array
	| tg.Mutation
	| tg.Template
	| Array<infer _U extends tg.Value>
	? T
	: T extends { [key: string]: tg.Value }
		? {
				[K in keyof T]?: T[K] | tg.Mutation<T[K]>;
			}
		: never;

export type MaybeReferent<T> = T | tg.Referent<T>;

export type UnresolvedArgs<T extends Array<tg.Value>> = {
	[K in keyof T]: tg.Unresolved<T[K]>;
};

export type ResolvedArgs<T extends Array<tg.Unresolved<tg.Value>>> = {
	[K in keyof T]: tg.Resolved<T[K]>;
};

export type ReturnValue = tg.MaybePromise<void> | tg.Unresolved<tg.Value>;

export type ResolvedReturnValue<T extends tg.ReturnValue> =
	T extends tg.MaybePromise<void>
		? undefined
		: T extends tg.Unresolved<tg.Value>
			? tg.Resolved<T>
			: never;
