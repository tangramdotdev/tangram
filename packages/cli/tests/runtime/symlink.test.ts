import { describe, expect, test } from "bun:test";
import Server from "../server.ts";
import { directory } from "../util.ts";

describe("constructor", () => {
	test("string", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const s = await tg.symlink("hello");
					return await s.object();
				});
			`,
		});
		const result = await server.tg`build ${dir}`.text().then((t) => t.trim());
		expect(result).toMatchSnapshot();
	});

	test("artifact", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const f = await tg.file({ contents: "hello" });
					const s = await tg.symlink(f);
					return await s.object();
				});
			`,
		});
		const result = await server.tg`build ${dir}`.text().then((t) => t.trim());
		expect(result).toMatchSnapshot();
	});

	test("object", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const f = await tg.file({ contents: "hello" });
					const d = await tg.directory({
						hello: tg.file("hello")
					});
					const s = await tg.symlink({ artifact: d, path: "hello" });
					return await s.object();
				});
			`,
		});
		const result = await server.tg`build ${dir}`.text().then((t) => t.trim());
		expect(result).toMatchSnapshot();
	});

	test("variadic overwrites completely", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const f = await tg.file({ contents: "hello" });
					const d1 = await tg.directory({
						hello: tg.file("hello from d1"),
					});
					const d2 = await tg.directory({
						hello: tg.file("hello from d2"),
					});
					const s = await tg.symlink({ artifact: d1, path: "hello" }, { artifact: d2 });
					return [await s.artifact(), await s.path()];
				});
			`,
		});
		const result = await server.tg`build ${dir}`.text().then((t) => t.trim());
		expect(result).toMatchSnapshot();
	});

	test("graph", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const d1 = await tg.directory({
						hello: tg.file("hello from d1"),
					});
					const graph = tg.graph({ nodes: [{ kind: "symlink", artifact: d1, path: "hello" }] });
					const s = await tg.symlink({ graph, node: 0 });
					return [await s.artifact(), await s.path()];
				});
			`,
		});
		const result = await server.tg`build ${dir}`.text().then((t) => t.trim());
		expect(result).toMatchSnapshot();
	});

	test("rejects mixed graph and regular objects", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
					export default tg.target(async () => {
					const d1 = await tg.directory({
						hello: tg.file("hello from d1"),
					});
					const graph = tg.graph({ nodes: [{ kind: "symlink", artifact: d1, path: "hello" }] });
					const s = await tg.symlink({ graph, node: 0 }, { path: "world" });
					return [await s.artifact(), await s.path()];
				});
			`,
		});
		const result = server.tg`build ${dir}`.quiet();
		await expect((async () => await result)()).rejects.toThrow();
	});

	test("rejects multiple graph objects", async () => {
		await using server = await Server.start();
		const dir = await directory({
			"tangram.ts": `
				export default tg.target(async () => {
					const d1 = await tg.directory({
						hello: tg.file("hello from d1"),
					});
					const graphA = tg.graph({ nodes: [{ kind: "symlink", artifact: d1, path: "hello" }] });
					let d2 = await tg.directory({
						hello: tg.file("hello from d2"),
					});
					const graphB = tg.graph({ nodes: [{ kind: "symlink", artifact: d2, path: "hello" }] });
					const s = await tg.symlink({ graph: graphA, node: 0 }, { graph: graphB, node: 0 });
					return [await s.artifact(), await s.path()];
				});
			`,
		});
		const result = server.tg`build ${dir}`.quiet();
		await expect((async () => await result)()).rejects.toThrow();
	});
});
