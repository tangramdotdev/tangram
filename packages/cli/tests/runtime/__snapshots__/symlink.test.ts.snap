// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor string 1`] = `"{"artifact":null,"path":"hello",}"`;

exports[`constructor artifact 1`] = `"{"artifact":fil_011ha7t8w6e20epbfm71yyaqzqgv7ypm24thgs2awqbj67fycra150,"path":null,}"`;

exports[`constructor object 1`] = `"{"artifact":dir_01vzsmc304fn7x92tzs48chcm35b6mtfgz3j9bgry8hd0y8h59gzgg,"path":"hello",}"`;

exports[`constructor variadic overwrites completely 1`] = `"[dir_01tr0h205a3bmedm8smte5j8mk8ewmm0wzf0pwrnk1nn3wcvevvpd0,"hello",]"`;

exports[`constructor graph 1`] = `"[dir_012er0x782axvp42narzyz58f81c474ktnsqak8r4cek3ech37cqq0,"hello",]"`;
